File: orchestrator/src/lifecycle/routing.ts
// Model routing bridge (B8/M4 fix) — TypeScript interface to the Python
// `rickgent_policies.select_model` router.
//
// The build loop calls `routeDispatch` before every dispatch. The router
// selects a harness/model/vendor from the live roster, enforcing:
//   - fail-closed on empty/unavailable roster (DENY, no dispatch)
//   - cross-vendor review (code_review role excludes implementer's vendor)
//   - cost gate (unpriced/over-hard-budget → DENY; over-soft-threshold → ASK)
//
// The call is a Python subprocess (`execFileSync("python3", ["-c", ...])`)
// that imports `select_model` and passes the roster + constraints as JSON.
// This is the SINGLE seam where the TS build path consults the Python router;
// the selected `vendor` flows into `Dispatcher.dispatch` as `opts.vendor` and
// from there into every ledger entry (B8 per-dispatch vendor label).

import { execFileSync } from "child_process";
import { RUNTIME_CAPABILITY_GATE } from "../capabilities/runtime-gate.js";

/** A model entry in the live roster (mirrors the Python `select_model` dict). */
export interface ModelEntry {
  harness: string;
  model: string;
  vendor: string;
  tier: "cheap" | "mid" | "capable";
  pricing: { cost_per_dispatch: number } | null;
}

/** A router selection (harness/model identity for one dispatch). */
export interface RouterSelection {
  harness: string;
  model: string;
  vendor: string;
}

/** The verdict returned by the router. */
export type RouterVerdict =
  | { result: "ALLOW"; selection: RouterSelection }
  | { result: "DENY"; reason: string; code: string }
  | { result: "ASK"; reason: string; code: string; selection: RouterSelection };

/** A non-ALLOW verdict (DENY or ASK) — the caller must NOT dispatch. */
export type RouterDenial =
  | { result: "DENY"; reason: string; code: string }
  | { result: "ASK"; reason: string; code: string; selection: RouterSelection };

/**
 * Call the Python `select_model` router via subprocess.
 *
 * Returns the parsed verdict dict. On subprocess failure, malformed output, or
 * timeout → fail-closed DENY (no dispatch, no silent fallback).
 */
export function callSelectModel(
  roster: ModelEntry[],
  role: string,
  implementerVendor?: string | null,
  costBudgetUsd?: number | null,
  softThresholdUsd?: number | null,
): RouterVerdict {
  const selectedPython = process.env["OMNIGENT_PYTHON"];
  if (selectedPython === undefined || selectedPython === "") {
    return {
      result: "DENY",
      reason: "routing: OMNIGENT_PYTHON is required; ambient python3 is forbidden",
      code: "ROUTING_PYTHON_NOT_SELECTED",
    };
  }
  const payload = JSON.stringify({
    roster,
    role,
    implementer_vendor: implementerVendor ?? null,
    cost_budget_usd: costBudgetUsd ?? null,
    soft_threshold_usd: softThresholdUsd ?? null,
  });

  const script =
    "import sys, json; " +
    "from rickgent_policies import select_model; " +
    "args = json.loads(sys.stdin.read()); " +
    "print(json.dumps(select_model(" +
    "args['roster'], args['role'], " +
    "implementer_vendor=args['implementer_vendor'], " +
    "cost_budget_usd=args['cost_budget_usd'], " +
    "soft_threshold_usd=args['soft_threshold_usd'])))";

  try {
    const stdout = execFileSync(selectedPython, ["-c", script], {
      encoding: "utf-8",
      input: payload,
      stdio: ["pipe", "pipe", "pipe"],
      timeout: 30000,
    });
    const parsed = JSON.parse(stdout.trim());
    if (parsed.result === "ALLOW" && parsed.selection) {
      return { result: "ALLOW", selection: parsed.selection };
    }
    if (parsed.result === "DENY") {
      return { result: "DENY", reason: parsed.reason ?? "routing denied", code: parsed.code ?? "ROUTING_DENIED" };
    }
    if (parsed.result === "ASK" && parsed.selection) {
      return { result: "ASK", reason: parsed.reason ?? "routing ask", code: parsed.code ?? "OVER_SOFT_THRESHOLD", selection: parsed.selection };
    }
    // Non-conforming return → fail closed.
    return { result: "DENY", reason: `routing: non-conforming router output: ${JSON.stringify(parsed)}`, code: "ROUTING_MALFORMED" };
  } catch (err) {
    return {
      result: "DENY",
      reason: `routing: subprocess error: ${err instanceof Error ? err.message : String(err)}`,
      code: "ROUTING_SUBPROCESS_ERROR",
    };
  }
}

/**
 * Route a dispatch: call the router and return either a go decision with the
 * selected vendor, or a no-go decision that the caller must NOT dispatch on.
 *
 * The caller (build loop) passes the selected `vendor` into
 * `Dispatcher.dispatch({vendor})` so it flows into the ledger. On DENY or ASK,
 * the caller records a fail-closed entry and absorbs via salvage/breaker.
 */
export function routeDispatch(
  roster: ModelEntry[],
  role: string,
  options?: {
    implementerVendor?: string | null;
    costBudgetUsd?: number | null;
    softThresholdUsd?: number | null;
  },
): { ok: true; selection: RouterSelection } | { ok: false; verdict: RouterDenial } {
  if (role === "code_review") RUNTIME_CAPABILITY_GATE.require("cross_vendor_review");
  const verdict = callSelectModel(
    roster,
    role,
    options?.implementerVendor ?? null,
    options?.costBudgetUsd ?? null,
    options?.softThresholdUsd ?? null,
  );
  if (verdict.result === "ALLOW") {
    return { ok: true, selection: verdict.selection };
  }
  return { ok: false, verdict: verdict as RouterDenial };
}

File: rickgent-policies/rickgent_policies/routing.py
"""Deterministic multi-vendor model routing."""

from __future__ import annotations

_ROLE_TIER_PREFERENCE = {
    "research": "cheap",
    "research_review": "cheap",
    "plan": "mid",
    "plan_review": "mid",
    "implement": "capable",
    "spec_conformance": "mid",
    "code_review": "mid",
    "simplify": "cheap",
}
_TIER_FALLBACK = ["cheap", "mid", "capable"]


def _tier_sort_key(model, preferred_tier):
    tier = model.get("tier", "mid")
    try:
        preferred = _TIER_FALLBACK.index(preferred_tier) if preferred_tier in _TIER_FALLBACK else 1
        actual = _TIER_FALLBACK.index(tier) if tier in _TIER_FALLBACK else 1
    except ValueError:
        preferred, actual = 1, 1
    return (abs(preferred - actual), actual, model.get("vendor", ""), model.get("model", ""))


def select_model(
    roster,
    role,
    implementer_vendor=None,
    cost_budget_usd=None,
    soft_threshold_usd=None,
):
    try:
        if not roster or not isinstance(roster, list):
            return {"result": "DENY", "reason": "routing: empty or unavailable roster", "code": "ROSTER_EMPTY"}
        candidates = list(roster)
        if role == "code_review" and implementer_vendor:
            candidates = [model for model in candidates if model.get("vendor") != implementer_vendor]
        if not candidates:
            return {
                "result": "DENY",
                "reason": f"routing: no candidates for role '{role}' after constraints",
                "code": "NO_CANDIDATES",
            }
        preferred_tier = _ROLE_TIER_PREFERENCE.get(role, "mid")
        candidates.sort(key=lambda model: _tier_sort_key(model, preferred_tier))
        ask_candidate = None
        for model in candidates:
            pricing = model.get("pricing")
            if not isinstance(pricing, dict):
                continue
            cost = pricing.get("cost_per_dispatch")
            if not isinstance(cost, (int, float)):
                continue
            if isinstance(cost_budget_usd, (int, float)) and cost > cost_budget_usd:
                continue
            if isinstance(soft_threshold_usd, (int, float)) and cost > soft_threshold_usd:
                ask_candidate = ask_candidate or model
                continue
            return {
                "result": "ALLOW",
                "selection": {key: model[key] for key in ("harness", "model", "vendor")},
            }
        if ask_candidate is not None:
            return {
                "result": "ASK",
                "reason": "routing: only over-soft-threshold candidates available",
                "code": "OVER_SOFT_THRESHOLD",
                "selection": {key: ask_candidate[key] for key in ("harness", "model", "vendor")},
            }
        return {
            "result": "DENY",
            "reason": "routing: no priced/within-budget model available",
            "code": "NO_PRICED_MODEL",
        }
    except Exception as error:
        return {"result": "DENY", "reason": f"routing: {error}", "code": "ROUTING_ERROR"}


__all__ = ["select_model"]
